packages/components/multiselect/src/Multiselect.tsx

```diff
@@ -0,0 +1,273 @@
+
+  const [searchValue, setSearchValue] = useState(defaultValue);
+  const [isOpen, setIsOpen] = useState(false);
+
+  const internalSearchInputRef = useRef(null);
+
+  const hasSearch = typeof onSearchValueChange === 'function';
+
+  const handleSearchChange = (event) => {
+    setSearchValue(event.target.value);
+    onSearchValueChange?.(event);
+  };
+
+  const resetSearch = () => {
+    // this looks a bit hacky, but is the official way of externally triggering the onChange handler for an input
+    // https://stackoverflow.com/a/46012210/17269164
+
+    const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
+      window.HTMLInputElement.prototype,
+      'value',
+    ).set;
+    nativeInputValueSetter.call(internalSearchInputRef.current, '');
+    const forcedEvent = new Event('change', { bubbles: true });
+    internalSearchInputRef.current.dispatchEvent(forcedEvent);
+  };
+
+  const renderMultiselectLabel = React.useCallback(() => {
+    if (currentSelection.length === 0) {
+      return <>{placeholder}</>;
+    }
+    const leftoverCount = currentSelection.length - 1;
+    if (leftoverCount === 0) {
+      return (
+        <span
+          data-test-id="cf-multiselect-current-selection"
+          className={styles.currentSelection}
+        >
+          {currentSelection[0]}
+        </span>
+      );
+    }
+    return (
+      <span
+        data-test-id="cf-multiselect-current-selection"
+        className={styles.currentSelection}
+      >
+        {currentSelection[0]} and {leftoverCount} more{' '}
+      </span>
+    );
+  }, [currentSelection, placeholder, styles.currentSelection]);
+
+  const childrenLength = React.Children.count(children);
+
+  return (
+    <div
+      data-test-id={testId}
+      className={cx(styles.multiselect, className)}
+      ref={ref}
+    >
+      <Popover
+        isOpen={isOpen}
+        onClose={() => setIsOpen(false)}
+        renderOnlyWhenOpen={false}
+        {...popoverProps}
+      >
+        <Popover.Trigger>
+          <Button
+            aria-label="Toggle Multiselect"
+            ref={toggleRef}
+            onClick={() => setIsOpen(!isOpen)}
+            startIcon={startIcon}
+            endIcon={<ChevronDownIcon />}
+          >
+            {renderMultiselectLabel()}
+          </Button>
+        </Popover.Trigger>
+        <Popover.Content
+          ref={listRef}
+          className={styles.content(listMaxHeight)}
+          testId="cf-multiselect-container"
+        >
+          <>
+            {hasSearch && (
+              <>
+                <TextInput
+                  aria-label="Search"
+                  type="text"
+                  value={searchValue}
+                  className={styles.inputField}
+                  id={id}
+                  testId="cf-multiselect-search"
+                  placeholder={searchPlaceholder}
+                  onChange={(event) => {
+                    handleSearchChange(event);
+                  }}
```

Comment:
nitpick: This means that `TextInput` will be re-rendered on every re-rendering of `Multiselect`. We could optimise this
```suggestion
                  onChange={handleSearchChange}
```
(And using useCallback for handleSearchChange)